### PR TITLE
Fix #311 by moving reports parameter in puppet.conf

### LINKED ISSUE #311: "reports" parameter needs to go into [main] section of puppet.conf
I was having problems with an instance of puppetserver deployed by theforeman-puppet displaying this error on agent run, when integrated with puppetDB:
```
[puppet-server] Puppet Attempt to assign to a reserved variable name: 'trusted' on node xmaster.vagrant.vm
```
I found the module is placing the "reports" parameter in the [master] section of the puppetserver's puppet.conf, when it appears to belong in the [main] section instead.
https://tickets.puppetlabs.com/browse/PDB-949?focusedCommentId=198934&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-198934

Also, acknowledging Puppet's own docs about this are vague:
https://docs.puppetlabs.com/guides/reporting.html

### PR DESCRIPTION
**Pull request by jyaworski**
Tag @mmoll and @westbywest

### CODE REVIEW

**Comment by mmoll** [comment]
Is this change compatible with 2.7 and 3.x? also, tests fail (as expected ;))

**Comment by mmoll** [comment]
@jyaworski ping?

**Comment by jyaworski** [comment]
@mmoll pong. I've been away on other tasks for a bit. I'll take another look at this.

**Comment by mmoll** [comment]
I tested this with a 2.7 and 3.8 client (with only the client installed, without any master) and with `puppet apply` this is leading to:
```
Warning: No report named 'foreman'
```
I can image this to happen even with puppet 4, so there has to get some logic in to set that parameter only when configuring a puppet server.

**Comment by jyaworski** [comment]
@mmoll I find it odd that this PR triggered that. I'll look into ways around it.

**Comment by mmoll** [comment]
I find it quite logical, actually. When in the main section, also the non-master install wants to send reports via foreman.rb, which is not installed on it, leading to that mesage. I'm a bit surprised PL did move the parameter, but I could imagine that this is on the other help helpful for masterless installs.

File: spec/classes/puppet_config_spec.rb [hunk 1]
```diff
@@ -11,6 +11,7 @@
           :domain                 => 'example.org',
           :fqdn                   => 'host.example.com',
           :puppetversion          => Puppet.version,
+          :server                 => true,
```
Comment (mmoll) [inline spec/classes/puppet_config_spec.rb]: what fact is this? :)

File: spec/classes/puppet_config_spec.rb [hunk 2]
```diff
@@ -282,6 +284,7 @@
       :osfamily => 'windows',
       :domain   => 'example.org',
       :puppetversion => Puppet.version,
+      :server => true,
```
Comment (mmoll) [inline spec/classes/puppet_config_spec.rb]: see above

File: spec/classes/puppet_server_config_spec.rb [hunk 3]
```diff
@@ -199,7 +199,7 @@
         end
 
         it 'should store reports' do
-          should contain_concat__fragment('puppet.conf+30-master').with_content(/^\s+reports\s+= store$/)
+          should contain_concat__fragment('puppet.conf+10-main').with_content(/^\s+reports\s+= store$/)
```
Comment (mmoll) [inline spec/classes/puppet_server_config_spec.rb]: this test can be deleted then, IMHO, as it's already tested in line 149.

File: spec/classes/puppet_config_spec.rb [hunk 4]
```diff
@@ -75,6 +76,7 @@
             '    hostprivkey = $privatekeydir/$certname.pem { mode = 640 }',
             '    autosign       = $confdir/autosign.conf { mode = 664 }',
             '    show_diff     = false',
+            '    reports     = foreman',
```
Comment (mmoll) [inline spec/classes/puppet_config_spec.rb]: this should not be there on agent only installs, correct?

File: spec/classes/puppet_config_spec.rb [hunk 5]
```diff
@@ -312,6 +315,7 @@
           '    hostprivkey = $privatekeydir/$certname.pem { mode = 640 }',
           '    autosign       = $confdir/autosign.conf { mode = 664 }',
           '    show_diff     = false',
+          '    reports     = foreman',
```
Comment (mmoll) [inline spec/classes/puppet_config_spec.rb]: see above

**Comment by jyaworski** [comment]
@mmoll derp. Fixed.

File: spec/classes/puppet_server_config_spec.rb [hunk 6]
```diff
@@ -10,9 +10,11 @@
           :clientcert             => 'puppetmaster.example.com',
           :concat_basedir         => '/nonexistant',
           :fqdn                   => 'puppetmaster.example.com',
-#          :rubyversion            => '1.9.3',
           :puppetversion          => Puppet.version,
       }) end
+      let (:params) {{
+          :server                 => true,
```
Comment (mmoll) [inline spec/classes/puppet_server_config_spec.rb]: that should be superflous, see line 68

**Comment by mmoll** [comment]
Tested with puppet 2.7, everything is still fine, so I think we're ready to go here. Please also add this test to make sure it's also working for the agent install:
```diff
diff --git a/spec/classes/puppet_config_spec.rb b/spec/classes/puppet_config_spec.rb
index a27b277..5e1bb47 100644
--- a/spec/classes/puppet_config_spec.rb
+++ b/spec/classes/puppet_config_spec.rb
@@ -64,6 +64,10 @@ describe 'puppet::config' do
           should_not contain_file('/etc/puppet/manifests/default_manifest.pp')
         end
 
+        it 'should_not contain reports setting in puppet.conf' do
+          should_not contain_concat__fragment('puppet.conf+10-main').with_content(/\s+reports = .*/)
+        end
+
         it 'should contain puppet.conf [main]' do
           concat_fragment_content = [
             '[main]',
```
Also squash the commits together into one and rebase onto current master.

**Comment by jyaworski** [comment]
@mmoll done.

**Comment by mmoll** [comment]
@jyaworski dzięki!
@domcleal @ekohl I think this is good to go. :+1:

**Comment by jyaworski** [comment]
@mmoll Proszę :)

**Comment by jyaworski** [comment]
So this works, but it looks like the tests timed out. We may want to consider addressing that project-wide.

**Comment by domcleal** [comment]
@jyaworski yes, we're tracking and slowly fixing individual causes under #330.

File: spec/classes/puppet_server_config_spec.rb [hunk 7]
```diff
@@ -10,7 +10,6 @@
           :clientcert             => 'puppetmaster.example.com',
           :concat_basedir         => '/nonexistant',
           :fqdn                   => 'puppetmaster.example.com',
-#          :rubyversion            => '1.9.3',
```
Comment (ekohl) [inline spec/classes/puppet_server_config_spec.rb]: I'd prefer to either remove this line or keep it in, but not comment it out.

Comment (mmoll) [inline spec/classes/puppet_server_config_spec.rb]: it's removed in that diff? :)

Comment (ekohl) [inline spec/classes/puppet_server_config_spec.rb]: And a good morning to me too ...

**Comment by ekohl** [comment]
The code looks good :+1:

**Comment by mmoll** [comment]
merged, jeszcze raz dzięki @jyaworski!

**Pull request merged**